### PR TITLE
Improve uninstall on Windows to remove directory even when uninstall is being executed from same directory

### LINKED ISSUE #2970: Uninstall on Windows triggers a Cloud IOC alert
The `uninstall` command for the Elastic Agent performs an action at the end of the installation process that spawns a `cmd.exe` that uses ping to wait 2 seconds and then delete the `C:\Program Files\Elastic\Agent` directory. This is done because `uninstall` is being executed from that directory and it cannot delete the directory that it is executing from (Windows does not allow it).

See the code here: https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/install/uninstall.go#L143

This type of process execution on Windows triggers a Cloud IOC alert. We should determine if its possible for this removal to be performed differently that doesn't trigger this type of alert.

- Version: All
- Operating System: Windows

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Previously the Elastic Agent would spawn a `cmd.exe` that would wait 2 seconds and then delete the `C:\Program Files\Elastic\Agent` directory. This triggers a Cloud IOC alert as that is something that nefarious applications do.

This provides a better way of uninstalling of Elastic Agent on Windows, by using the Windows API to allow removal of a directory even if there is an executable running inside of it.

## Why is it important?

Two reasons this is important. Uninstall no longer has a 2 second delay for it to be fully removed, something that caused a problem with integration testing on Windows, and a Cloud IOC alert will no longer trigger when Elastic Agent is uninstalled.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test (enabled with Windows runner in other PR)

## How to test this PR locally

```
PS> .\elastic-agent install -f
PS> "C:\Program Files\Elastic\Agent\elastic-agent.exe" uninstall -f
```

## Related issues

- Closes #2970 

## Logs

```
PS C:\Program Files\Elastic> .\Agent\elastic-agent.exe uninstall
Elastic Agent will be uninstalled from your system at C:\Program Files\Elastic\Agent. Do you want to continue? [Y/n]:y
Elastic Agent has been uninstalled.
```

### CODE REVIEW

File: internal/pkg/agent/install/uninstall_windows.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		return "", fmt.Errorf("failed to open handler for %q: %w", path, err)
+	}
+
+	// rename handle
+	err = renameHandle(h)
+	if err != nil {
+		_ = windows.CloseHandle(h)
+		return "", fmt.Errorf("failed to rename handler for %q: %w", path, err)
+	}
+	_ = windows.CloseHandle(h)
```
Comment (cmacknz) [inline internal/pkg/agent/install/uninstall_windows.go:45]:
```suggestion
	err = renameHandle(h)
	_ = windows.CloseHandle(h)
	if err != nil {
		return "", fmt.Errorf("failed to rename handler for %q: %w", path, err)
	}
```

Looks like you could just unconditionally close the handler here.

File: internal/pkg/agent/install/uninstall_windows.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+		return "", fmt.Errorf("failed to open handler after rename for %q: %w", path, err)
+	}
+
+	// disposite the handle
+	err = depositeHandle(h)
+	if err != nil {
+		_ = windows.CloseHandle(h)
+		return "", fmt.Errorf("failed to deposite handler for %q: %w", path, err)
+	}
+	_ = windows.CloseHandle(h)
```
Comment (cmacknz) [inline internal/pkg/agent/install/uninstall_windows.go:59]:
```suggestion
	err = disposeHandle(h)
	_ = windows.CloseHandle(h)
	if err != nil {
		return "", fmt.Errorf("failed to dispose handler for %q: %w", path, err)
	}
```

Should be `dispose` in the name going off of the `FileDispositionInfo` later on. Also looks like we can unconditionally close the handle before checking the error.

File: internal/pkg/agent/install/uninstall_windows.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+func removeBlockingExe(blockingErr error) (string, error) {
+	path := getPathFromError(blockingErr)
+	if path == "" {
+		return "", nil
+	}
+
+	// open handle for delete only
+	h, err := openDeleteHandle(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open handler for %q: %w", path, err)
```
Comment (ycombinator) [inline internal/pkg/agent/install/uninstall_windows.go:36]:
```suggestion
		return "", fmt.Errorf("failed to open handle for %q: %w", path, err)
```

File: internal/pkg/agent/install/uninstall_windows.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	// rename handle
+	err = renameHandle(h)
+	if err != nil {
+		_ = windows.CloseHandle(h)
+		return "", fmt.Errorf("failed to rename handler for %q: %w", path, err)
```
Comment (ycombinator) [inline internal/pkg/agent/install/uninstall_windows.go:43]:
```suggestion
		return "", fmt.Errorf("failed to rename handle for %q: %w", path, err)
```

File: internal/pkg/agent/install/uninstall_windows.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	_ = windows.CloseHandle(h)
+
+	// re-open handle
+	h, err = openDeleteHandle(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open handler after rename for %q: %w", path, err)
```
Comment (ycombinator) [inline internal/pkg/agent/install/uninstall_windows.go:50]:
```suggestion
		return "", fmt.Errorf("failed to open handle after rename for %q: %w", path, err)
```

File: internal/pkg/agent/install/uninstall_windows_test.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+	"github.com/stretchr/testify/require"
+)
+
+const simpleBlockForever = `
+package main
+func main() {
+    select{}
+}
+`
```
Comment (ycombinator) [inline internal/pkg/agent/install/uninstall_windows_test.go:23]: When I try to run this program (admittedly on my Mac), it exits pretty much immediately citing a deadlock:

```
$ time go run block.go
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [select (no cases)]:
main.main()
	/tmp/block.go:4 +0x1c
exit status 2
go run block.go  0.05s user 0.06s system 59% cpu 0.193 total
```

Is it possible the same thing could happen in your test before `RemovePath` gets called?

Comment (blakerouse) [reply internal/pkg/agent/install/uninstall_windows_test.go]: I ran the tests with the debugger to validate that I was getting the error handling correct and that the code actually works, using the same code.

But I do see that same error, so I have update the test to sleep forever.

**Review by ycombinator — APPROVED** [review]
LGTM.

File: internal/pkg/agent/install/uninstall_windows.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		windows.FILE_ATTRIBUTE_NORMAL,
+		0,
+	)
+	if err != nil {
+		return 0, err
+	}
+	return handle, nil
+}
+
+func renameHandle(hHandle windows.Handle) error {
```
Comment (leehinman) [inline internal/pkg/agent/install/uninstall_windows.go:94]: is there any danger to running removeBlockingExe twice in a row?  It looks like we always rename to the same name, so if setting the fileDispositionInfo to DeleteFile requires some kind of action to happen before it is called again we might have a weird race condition.

Maybe a fixed prefix with some random characters would be better as a filename?

**Review by leehinman — COMMENTED** [review]
Sorry for not getting to this yesterday.

Any chance you could document the logic behind what is going on deleteHandle and changing the file disposition?  If something changes with Windows filesystems 5 years from now a little extra context will go a long way.

**Pull request merged**